### PR TITLE
Add some three-argument ldiv! methods

### LINKED ISSUE #615: ldiv!, rdiv!, and missing methods
(this is three issues in one, I can split if preferred; I just want to get feedback on whether the current behavior is by design or an oversight)

- `ldiv!` has a `ldiv!(Y,A,B)` form which `rdiv!` doesn't.

- There appears to be a number of missing methods, such as with `Diagonal`.

- I find it strange that `ldiv!(A,A)` when `A` is a matrix doesn't work while `A\A` does. I understand the rationale explained in the docstring, but presumably if an user is advanced enough to go to `ldiv!` they also know that factorizations are expensive and require memory. Not having this means that, for instance, a library that wants to support preconditioning can not just call `ldiv!` and expect to work on a wide range of types, and be performant when necessary.

### PR DESCRIPTION
**Pull request by antoine-levitt**
Fix https://github.com/JuliaLang/LinearAlgebra.jl/issues/615, to make honest citizens out of NLsolve and DiffEqBase (cc @pkofod @ChrisRackauckas)

### CODE REVIEW

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -427,7 +427,7 @@ function ldiv!(D::Diagonal{T}, V::AbstractMatrix{T}) where {T}
     end
     V
 end
-
+LinearAlgebra.ldiv!(x::AbstractArray,A::Diagonal,b::AbstractArray) = (x .= A.diag .\ b)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/diagonal.jl]: * there is no need for the specifier `LinearAlgebra`
* there should be a space after each comma in the list of arguments
* should we be a bit more specific with regards to `x` and `b`? Maybe `AbstractVecOrMat`? not sure
* while you're adding `ldiv!`, let's have `rdiv!` as well!

File: stdlib/LinearAlgebra/src/generic.jl [hunk 2]
```diff
@@ -238,6 +238,11 @@ function ldiv!(s::Number, X::AbstractArray)
     end
     X
 end
+ldiv!(Y::AbstractArray, s::Number, X::AbstractArray) = Y .= s .\ X
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/generic.jl]: Add `rdiv!`?

File: stdlib/LinearAlgebra/src/generic.jl [hunk 3]
```diff
@@ -238,6 +238,11 @@ function ldiv!(s::Number, X::AbstractArray)
...
+
+# Generic fallback. This assumes that B and Y have the same sizes.
+LinearAlgebra.ldiv!(Y::AbstractArray, A::AbstractMatrix, B::AbstractArray) = ldiv!(A,copyto!(Y,B))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/generic.jl]: unnecessary specifier and spaces

**Review by dkarrasch — CHANGES_REQUESTED** [review]
Amazing that this doesn't introduce ambiguities! I guess my comments extend to all changes:
* add `rdiv!` while you're at it
* `AbstractArray` may be too broad of an argument type
* and some stylistic remarks

**Comment by antoine-levitt** [comment]
Thanks! Adressed the style issues. Regarding the types, I really don't know. It seemed to me that having the fallbacks be the broadest possible would avoid ambiguities, but I'm not very familiar with that. Regarding `rdiv!`, @andreasnoack was against in https://github.com/JuliaLang/julia/issues/31278, on the grounds that it'd just be misleading users since most types don't support efficient `rdiv!`. These are very low-level methods and `ldiv!` is much more common, so I think omitting it is fine, but I'm happy to add it if there's a consensus.

File: stdlib/LinearAlgebra/test/cholesky.jl [hunk 4]
```diff
@@ -180,6 +180,7 @@ end
                 # Test error bound on linear solver: LAWNS 14, Theorem 2.1
                 # This is a surprisingly loose bound
                 BB = copy(B)
+                out = zero(BB)
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/test/cholesky.jl]: What is this line good for? Seems like you haven't made any changes to the cholesky code.

**Review by dkarrasch — APPROVED** [review]

File: stdlib/LinearAlgebra/test/cholesky.jl (hunk 4 above)
Comment (antoine-levitt) [reply stdlib/LinearAlgebra/test/cholesky.jl]: Just me being stupid, sorry

**Comment by dkarrasch** [comment]
One more thing that came to my mind. If this is a new method (with that signature), then we should perhaps announced in `NEWS.md`, so that (honest 😛) package developers will get a chance to know that they can either rely on the `Base` version or overload the function. In fact, they may run into warnings because of overloading an existing function, right?

**Comment by antoine-levitt** [comment]
I think this falls under the category of "bugfix" and does not need news or compat, but I'm happy to do it if you feel it's appropriate.

**Comment by antoine-levitt** [comment]
Anything that needs to be done here?

**Comment by lkapelevich** [comment]
Would love to be able to use this!

**Comment by chriscoey** [comment]
Ditto, bump!

**Comment by ChrisRackauckas** [comment]
☠️ check matey! ☠️

**Pull request merged**